aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp

```diff
@@ -204,17 +209,186 @@ class QLinearDynamicInt8 final : public torch::OperatorKernel {
 
     return output;
   }
-#else // USE_FBGEMM
-  at::Tensor operator()(
-      at::Tensor /* input */,
-      at::Tensor /* packed_weight */) {
-    // We make a strong guarantee that models using these operators will have
-    // the same numerics across different machines. Therefore, we do not provide
-    // a fallback path and rather fail loudly if we cannot run FBGEMM.
+#endif // USE_FBGEMM
+#ifdef USE_PYTORCH_QNNPACK
+
+  struct qparams {
+    float scale;
+    std::int32_t zero_point;
+  };
+
+  qparams chooseQuantizationParams(
+      float min,
+      float max,
+      int32_t qmin,
+      int32_t qmax) {
+    double scale =
+        (std::max(max, 0.f) - std::min(min, 0.f)) / ((double)qmax - qmin);
+    if (scale == 0) {
+      scale = 0.1;
+    }
+    // If scale is 0, we arbitrary adjust the scale to 0.1
+    assert(scale > 0);
+    // We extend the [min, max] interval to ensure that it contains 0.
+    // Otherwise, we would not meet the requirement that 0 be an exactly
+    // representable value.
+    min = std::min(min, 0.f);
+    max = std::max(max, 0.f);
+
+    double zero_point_from_min = qmin - min / scale;
+    double zero_point_from_max = qmax - max / scale;
+    double zero_point_from_min_error = std::abs(qmin) + std::abs(min / scale);
+    double zero_point_from_max_error = std::abs(qmax) + std::abs(max / scale);
+    double initial_zero_point =
+        zero_point_from_min_error < zero_point_from_max_error
+        ? zero_point_from_min
+        : zero_point_from_max;
```

Comment:
Did you mean to do

```c++
double zero_point_from_min_error = std::abs(qmin) - std::abs(min / scale);
double zero_point_from_max_error = std::abs(qmax) - std::abs(max / scale);
```

If not, this is not really an absolute error, and `initial_zero_point` will always be equal to `zero_point_from_max`

Reply:
The arithmetic error on the zero point computed from min or max will depend on the value of individual terms. It will roughly be epsilon * (sum of absolute values of terms).
(This is inspired from what FBGEMM does today, as I didn't want the calculations to diverge)

Reply:
Here is the reference implementation used for static quant (doesn't need to align with this implementation though)
```
                scale = (max_val - min_val) / float(qmax - qmin)
                scale = max(scale, self.eps)
                zero_point = qmin - round(min_val / scale)
                zero_point = max(qmin, zero_point)
                zero_point = min(qmax, zero_point)
                zero_point = int(zero_point)
```

Reply:
> The arithmetic error on the zero point computed from min or max will depend on the value of individual terms. It will roughly be epsilon * (sum of absolute values of terms).
> (This is inspired from what FBGEMM does today, as I didn't want the calculations to diverge)

I understand that you don't want to diverge from FBGEMM, but these lines don't make sense -- because `qmax` is guaranteed to be greater than `qmin`, and `max` is guarantted to be greater or equal than `min`, `initial_zero_point` will always be `zero_point_from_max_error`. I am almost positive FBGEMM meant to do `std::abs(qmin) - std::abs(min / scale)` instead of `+`. It is better to send a fix suggestion to FBGEMM instead of propagating the bug here.

Reply:
cc @jianyuh to the discussion